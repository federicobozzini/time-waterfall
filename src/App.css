@font-face{
  font-family:'digital-clock-font';
  src: url('fonts/digital-7.ttf');
 }

body {
  width: 100%;
  height: 100vh;
  margin: 0px;
  cursor: none;
}

#root {
  background: #111;
  height: 100%;
  width: 100%;
}

#app {
  position:fixed;
  top:0;
  left:0;
  height:100%;
  width:100%
}

text {
  fill: white;
  font-family: digital-clock-font;
  transform-box: fill-box;
  transform-origin: center;
}
