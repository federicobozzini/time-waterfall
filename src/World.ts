import { Timer } from "./Timer";

type OnElementsUpdate = (elements: Element[]) => void;

export interface Element {
    x: number;
    y: number;
    angle: number;
    value: number; /* 1 to 9 */
    size: number;
}


const l0 = {
    x: 95, 
    y: 40,
    angle: -30,
    value: 8,
    size: 12,
};
const l1 = {
    x: 40, 
    y: 90,
    angle: 5,
    value: 4,
    size: 8,
};
const l2 = {
    x: 20, 
    y: 15,
    angle: -70,
    value: 3,
    size: 22,
};
const l3 = {
    x: 85, 
    y: 75,
    angle: 10,
    value: 1,
    size: 8,
};

const yVel = 10;
const yMax = 130;

export class World {

    private elements = [l0, l1, l2, l3];
    private updateCbs: OnElementsUpdate[] = [];

    constructor(timer: Timer) {
        timer.onTick((interval: number) => {
            this.update(interval);
        });
    }

    public onUpdate(cb: OnElementsUpdate) {
        this.updateCbs.push(cb);
    }
    
    public getElements(): Element[] {
        return this.elements;
    }

    public removeCb(cb: OnElementsUpdate) {
        this.updateCbs = this.updateCbs.filter(cb1 => cb1 !== cb);
    }

    private update(interval:number) {
        this.elements = this.elements.map(e => 
            ({
                ...e, 
                y: e.y + yVel * (interval / 1000)
            })).filter(e => e.y < yMax);
        this.updateCbs.forEach(cb => cb(this.elements));
    }

}