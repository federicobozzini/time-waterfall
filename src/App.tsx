import React from 'react'
import './App.css';

const App: React.FC = () => {
  return (
    <svg viewBox="0 0 100 100" preserveAspectRatio="none" xmlns="http://www.w3.org/2000/svg" id='app'>
      <text x="95" y="40" fontSize="12" transform="rotate(-30, 95, 40)">8</text>
      <text x="40" y="90" fontSize="8" transform="rotate(5, 40, 90)">4</text>
      <text x="20" y="15" fontSize="22" transform="rotate(-70, 20, 15)">3</text>
      <text x="85" y="75" fontSize="8" transform="rotate(10, 85, 75)">1</text>
    </svg>
  );
}

export default App;
