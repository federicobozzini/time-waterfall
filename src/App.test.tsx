import React from 'react';
import { render } from '@testing-library/react';
import App from './App';
import { World } from './World';


const world = new World();

describe('App rendering', () => {

  test('renders an element', () => {
    const { getByText } = render(<App world={world} />);
    const numberElement = getByText(/8/i);
    expect(numberElement).toBeInTheDocument();
  });

});
