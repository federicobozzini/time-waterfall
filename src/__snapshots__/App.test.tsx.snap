// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App rendering renders the full app 1`] = `
<div>
  <svg
    id="app"
    preserveAspectRatio="none"
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  />
</div>
`;

exports[`App rendering renders the full app 2`] = `
<div>
  <svg
    id="app"
    preserveAspectRatio="none"
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <text
      font-size="15.289891721215099"
      transform="rotate(-165.21320178285245)"
      x="53.77714571157936"
      y="1.362047513346188"
    >
      3
    </text>
  </svg>
</div>
`;

exports[`App rendering renders the full app 3`] = `
<div>
  <svg
    id="app"
    preserveAspectRatio="none"
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <text
      font-size="15.289891721215099"
      transform="rotate(-218.4547174640006)"
      x="53.37225356638894"
      y="20.819869132577413"
    >
      2
    </text>
    <text
      font-size="15.149289197288454"
      transform="rotate(-40.0234843287617)"
      x="34.66990695750335"
      y="42.14579240852033"
    >
      2
    </text>
    <text
      font-size="6.892302706371993"
      transform="rotate(-156.6810995324517)"
      x="64.99528671470003"
      y="8.247276416961103"
    >
      1
    </text>
  </svg>
</div>
`;

exports[`App rendering renders the full app 4`] = `
<div>
  <svg
    id="app"
    preserveAspectRatio="none"
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <text
      font-size="6.282950809691101"
      transform="rotate(-266.6746498388238)"
      x="79.62382787028824"
      y="79.85008671564533"
    >
      2
    </text>
    <text
      font-size="16.044193344656378"
      transform="rotate(-26.967077577486634)"
      x="56.27094857295972"
      y="13.524224523184815"
    >
      1
    </text>
  </svg>
</div>
`;

exports[`App rendering renders the full app 5`] = `
<div>
  <svg
    id="app"
    preserveAspectRatio="none"
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <text
      font-size="15.918192766606808"
      transform="rotate(100.571814166009)"
      x="66.36058797982028"
      y="81.11453728459824"
    >
      2
    </text>
    <text
      font-size="21.78380083758384"
      transform="rotate(-86.57665416151308)"
      x="77.82247395560259"
      y="62.707205129884755"
    >
      3
    </text>
    <text
      font-size="6.87708405777812"
      transform="rotate(-95.25253353640417)"
      x="35.46382457204176"
      y="40.98408845392979"
    >
      1
    </text>
    <text
      font-size="12.848742100875825"
      transform="rotate(-197.16910644136314)"
      x="54.45369487096592"
      y="99.45110984149848"
    >
      4
    </text>
    <text
      font-size="15.791896470822394"
      transform="rotate(-67.77316603325303)"
      x="82.38758779662635"
      y="36.28338720353324"
    >
      5
    </text>
    <text
      font-size="7.657657964620739"
      transform="rotate(138.3397817540915)"
      x="49.20257927967697"
      y="13.475160997824275"
    >
      7
    </text>
    <text
      font-size="22.53528465796262"
      transform="rotate(131.42308366494245)"
      x="15.561488375919867"
      y="1.0834121309872715"
    >
      8
    </text>
    <text
      font-size="13.359900433570147"
      transform="rotate(72.47431498160586)"
      x="99.70979543026537"
      y="0.3953779986617155"
    >
      5
    </text>
  </svg>
</div>
`;

exports[`App rendering renders the full app 6`] = `
<div>
  <svg
    id="app"
    preserveAspectRatio="none"
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <text
      font-size="15.429372511804104"
      transform="rotate(262.47897048164026)"
      x="33.935035647618534"
      y="95.12327731770891"
    >
      2
    </text>
    <text
      font-size="22.513529579155147"
      transform="rotate(363.1510221194476)"
      x="103.84040494558784"
      y="62.999912713018816"
    >
      3
    </text>
    <text
      font-size="11.20789027120918"
      transform="rotate(28.367694643512653)"
      x="52.371188804656526"
      y="21.2010964295827"
    >
      4
    </text>
    <text
      font-size="11.338602618779987"
      transform="rotate(25.6540492060596)"
      x="84.86309334356363"
      y="50.66551247471938"
    >
      8
    </text>
    <text
      font-size="15.087324963882565"
      transform="rotate(-178.6786413654686)"
      x="30.515126408934606"
      y="5.962578777037562"
    >
      7
    </text>
    <text
      font-size="19.10554161434993"
      transform="rotate(-156.01511211600155)"
      x="37.84743474163115"
      y="0.794018121866975"
    >
      7
    </text>
  </svg>
</div>
`;

exports[`App rendering renders the full app 7`] = `
<div>
  <svg
    id="app"
    preserveAspectRatio="none"
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <text
      font-size="19.05685656191781"
      transform="rotate(74.04798576179891)"
      x="6.821552326441768"
      y="20.203549530354785"
    >
      7
    </text>
    <text
      font-size="22.669884414877743"
      transform="rotate(180.5991956569249)"
      x="17.708461387693024"
      y="10.977077074642304"
    >
      7
    </text>
    <text
      font-size="7.158290677238256"
      transform="rotate(58.113469601422594)"
      x="91.39430201895539"
      y="10.04816882561892"
    >
      8
    </text>
    <text
      font-size="7.800996467471123"
      transform="rotate(113.22235981654373)"
      x="32.731386246765055"
      y="3.761392316431738"
    >
      4
    </text>
  </svg>
</div>
`;
